// @kaleidoswap/kaleidoswap-sdk/dist/http/client.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.HttpClient = void 0;
const errorFactory_1 = require("../types/errorFactory");
const exceptions_1 = require("../types/exceptions");
const retry_1 = require("../utils/retry");
/**
 * HTTP client for making requests to the KaleidoSwap API
 */
class HttpClient {
    /**
     * Makes a GET request
     * @param endpoint - API endpoint
     * @returns Promise resolving to the response data
     */
    async get(endpoint) {
        return this.request('GET', endpoint);
    }
    /**
     * Makes a POST request
     * @param endpoint - API endpoint
     * @param data - Request data
     * @returns Promise resolving to the response data
     */
    async post(endpoint, data) {
        return this.request('POST', endpoint, data);
    }
    /**
     * Creates a new HTTP client instance
     * @param config - Configuration for the client
     */
    constructor(config) {
        this.baseUrl = config.baseUrl.replace(/\/$/, '');
        this.apiKey = config.apiKey;
        this.retryConfig = config.retryConfig || {};
    }
    /**
     * Gets the headers for API requests
     * @returns Object containing request headers
     */
    getHeaders() {
        const headers = {
            'Content-Type': 'application/json',
            'Accept': 'application/json'
        };
        if (this.apiKey) {
            headers['Authorization'] = `Bearer ${this.apiKey}`;
        }
        return headers;
    }
    /**
     * Makes an HTTP request with retry logic
     * @param method - HTTP method (GET, POST, etc.)
     * @param endpoint - API endpoint
     * @param data - Optional request data
     * @returns Promise resolving to the response data
     * @throws {AuthenticationError} If authentication fails
     * @throws {RateLimitError} If rate limit is exceeded
     * @throws {ValidationError} If request validation fails
     * @throws {NetworkError} If there's a network error
     * @throws {TimeoutError} If request times out
     */
    async request(method, endpoint, data) {
        return (0, retry_1.retry)(async () => {
            const url = `${this.baseUrl}${endpoint}`;
            const headers = this.getHeaders();
            const requestId = `req_${Date.now()}_${Math.random().toString(36).substr(2, 9)}`;
            const options = {
                method,
                headers: {
                    ...headers,
                    'Content-Type': 'application/json',
                    'X-Request-ID': requestId,
                },
                body: data ? JSON.stringify(data) : undefined,
            };
            try {
                const controller = new AbortController();
                const timeoutId = setTimeout(() => controller.abort(), 30000); // 30 second timeout
                const response = await fetch(url, {
                    ...options,
                    signal: controller.signal
                });
                clearTimeout(timeoutId);
                const responseText = await response.text();
                if (!response.ok) {
                    throw errorFactory_1.ErrorFactory.fromHttpResponse(response, responseText, data, requestId);
                }
                if (!responseText) {
                    return {};
                }
                try {
                    return JSON.parse(responseText);
                }
                catch (parseError) {
                    throw errorFactory_1.ErrorFactory.createValidationError('response', responseText, 'Invalid JSON format in server response', { endpoint, method });
                }
            }
            catch (error) {
                // Re-throw SDK errors as-is
                if (error instanceof exceptions_1.AuthenticationError ||
                    error instanceof exceptions_1.RateLimitError ||
                    error instanceof exceptions_1.ValidationError ||
                    error instanceof exceptions_1.NetworkError ||
                    error instanceof exceptions_1.TimeoutError) {
                    throw error;
                }
                // Handle AbortError (timeout)
                if (error instanceof Error && error.name === 'AbortError') {
                    throw errorFactory_1.ErrorFactory.createTimeoutError(`${method} ${endpoint}`, 30000, { data, requestId });
                }
                // Handle other network errors
                throw errorFactory_1.ErrorFactory.fromNetworkError(error instanceof Error ? error : new Error(String(error)), data, requestId);
            }
        }, this.retryConfig);
    }
}
exports.HttpClient = HttpClient;
