// @kaleidoswap/kaleidoswap-sdk/dist/types/errorFactory.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ErrorFactory = void 0;
const errors_1 = require("./errors");
const exceptions_1 = require("./exceptions");
/**
 * Factory for creating appropriate error instances from different contexts
 */
class ErrorFactory {
    /**
     * Create an error from an HTTP response
     */
    static fromHttpResponse(response, responseText, requestData, requestId) {
        const statusCode = response.status;
        const metadata = {
            statusCode,
            response: responseText,
            requestData,
            requestId,
            timestamp: new Date()
        };
        let message = `HTTP ${statusCode}: ${response.statusText}`;
        if (responseText) {
            try {
                const parsed = JSON.parse(responseText);
                message = parsed.message || parsed.error || message;
            }
            catch {
                // Keep original message if JSON parsing fails
            }
        }
        // Map specific HTTP status codes to appropriate error types
        switch (statusCode) {
            case 401:
                return new exceptions_1.AuthenticationError(message, errors_1.ErrorCode.AUTH_INVALID_API_KEY, metadata);
            case 403:
                return new exceptions_1.AuthenticationError(message, errors_1.ErrorCode.AUTH_INSUFFICIENT_PERMISSIONS, metadata);
            case 429:
                return new exceptions_1.RateLimitError(message, errors_1.ErrorCode.RATE_LIMIT_EXCEEDED, metadata);
            case 400:
            case 422:
                return new exceptions_1.ValidationError(message, errors_1.ErrorCode.VALIDATION_INVALID_FORMAT, metadata);
            default:
                return new exceptions_1.HttpError(message, statusCode, undefined, metadata);
        }
    }
    /**
     * Create an error from a fetch/network error
     */
    static fromNetworkError(error, requestData, requestId) {
        const metadata = {
            requestData,
            requestId,
            timestamp: new Date(),
            retryable: true,
            retryStrategy: errors_1.RetryStrategy.EXPONENTIAL_BACKOFF
        };
        let code = errors_1.ErrorCode.NETWORK_UNREACHABLE;
        let message = error.message;
        // Map specific network errors to error codes
        if (error.message.includes('timeout')) {
            code = errors_1.ErrorCode.CONNECTION_TIMEOUT;
            message = 'Request timed out. Please try again.';
        }
        else if (error.message.includes('ECONNREFUSED')) {
            code = errors_1.ErrorCode.CONNECTION_REFUSED;
            message = 'Connection refused by server.';
        }
        else if (error.message.includes('ENOTFOUND')) {
            code = errors_1.ErrorCode.DNS_RESOLUTION_FAILED;
            message = 'Could not resolve server address.';
        }
        else if (error.name === 'AbortError') {
            code = errors_1.ErrorCode.OPERATION_CANCELLED;
            message = 'Request was cancelled.';
        }
        return new exceptions_1.NetworkError(message, code, metadata, error);
    }
    /**
     * Create an error from a WebSocket error
     */
    static fromWebSocketError(error, context) {
        const metadata = {
            context: { websocketContext: context },
            timestamp: new Date(),
            retryable: true,
            retryStrategy: errors_1.RetryStrategy.EXPONENTIAL_BACKOFF,
            retryDelay: 1000,
            maxRetries: 5
        };
        let code = errors_1.ErrorCode.WEBSOCKET_CONNECTION_FAILED;
        let message = 'WebSocket connection failed';
        if (error instanceof Error) {
            message = error.message;
            if (error.message.includes('close')) {
                code = errors_1.ErrorCode.WEBSOCKET_DISCONNECTED;
                message = 'WebSocket connection closed unexpectedly';
            }
        }
        else if (error instanceof Event) {
            if (error.type === 'close') {
                code = errors_1.ErrorCode.WEBSOCKET_DISCONNECTED;
                message = 'WebSocket connection closed';
            }
        }
        return new exceptions_1.WebSocketError(message, code, metadata, error instanceof Error ? error : undefined);
    }
    /**
     * Create a validation error with specific field information
     */
    static createValidationError(field, value, reason, requestData) {
        const message = `Validation failed for field '${field}': ${reason}`;
        const metadata = {
            context: {
                field,
                value,
                reason,
                requestData
            },
            timestamp: new Date(),
            retryable: false,
            recoveryActions: [
                `Check the '${field}' field value`,
                'Refer to API documentation for valid formats',
                'Contact support if the issue persists'
            ]
        };
        let code = errors_1.ErrorCode.VALIDATION_INVALID_FORMAT;
        if (reason.includes('required') || reason.includes('missing')) {
            code = errors_1.ErrorCode.VALIDATION_REQUIRED_FIELD_MISSING;
        }
        else if (reason.includes('range') || reason.includes('bounds')) {
            code = errors_1.ErrorCode.VALIDATION_OUT_OF_RANGE;
        }
        return new exceptions_1.ValidationError(message, code, metadata);
    }
    /**
     * Create an asset-related error
     */
    static createAssetError(assetId, operation, reason) {
        const message = `Asset error for '${assetId}' during ${operation}${reason ? ': ' + reason : ''}`;
        const metadata = {
            context: {
                assetId,
                operation,
                reason
            },
            timestamp: new Date(),
            recoveryActions: [
                'Verify the asset ID is correct',
                'Check if the asset is supported',
                'Try refreshing asset data'
            ]
        };
        let code = errors_1.ErrorCode.ASSET_NOT_FOUND;
        if (reason?.includes('not supported')) {
            code = errors_1.ErrorCode.ASSET_NOT_SUPPORTED;
        }
        return new exceptions_1.AssetError(message, code, metadata);
    }
    /**
     * Create a trading pair error
     */
    static createPairError(baseAsset, quoteAsset, operation, reason) {
        const pairName = `${baseAsset}/${quoteAsset}`;
        const message = `Pair error for '${pairName}' during ${operation}${reason ? ': ' + reason : ''}`;
        const metadata = {
            context: {
                baseAsset,
                quoteAsset,
                pairName,
                operation,
                reason
            },
            timestamp: new Date(),
            recoveryActions: [
                'Verify both asset IDs are correct',
                'Check if the trading pair is supported',
                'Try a different trading pair'
            ]
        };
        let code = errors_1.ErrorCode.PAIR_NOT_FOUND;
        if (reason?.includes('not supported')) {
            code = errors_1.ErrorCode.PAIR_NOT_SUPPORTED;
        }
        return new exceptions_1.PairError(message, code, metadata);
    }
    /**
     * Create a quote-related error
     */
    static createQuoteError(quoteId, reason) {
        const message = quoteId
            ? `Quote error for '${quoteId}': ${reason}`
            : `Quote error: ${reason}`;
        const metadata = {
            context: {
                quoteId,
                reason
            },
            timestamp: new Date(),
            recoveryActions: [
                'Request a new quote',
                'Check if the quote parameters are valid',
                'Verify sufficient liquidity exists'
            ]
        };
        let code = errors_1.ErrorCode.QUOTE_NOT_FOUND;
        if (reason.includes('expired')) {
            code = errors_1.ErrorCode.QUOTE_EXPIRED;
        }
        return new exceptions_1.QuoteError(message, code, metadata);
    }
    /**
     * Create a swap-related error
     */
    static createSwapError(operation, reason, swapId, additionalContext) {
        const message = swapId
            ? `Swap ${operation} failed for '${swapId}': ${reason}`
            : `Swap ${operation} failed: ${reason}`;
        const metadata = {
            context: {
                operation,
                reason,
                swapId,
                ...additionalContext
            },
            timestamp: new Date(),
            recoveryActions: [
                'Check swap parameters',
                'Verify sufficient balance',
                'Request a new quote if expired',
                'Contact support if issue persists'
            ]
        };
        let code = errors_1.ErrorCode.SWAP_FAILED;
        if (reason.includes('timeout')) {
            code = errors_1.ErrorCode.SWAP_TIMEOUT;
        }
        else if (reason.includes('balance') || reason.includes('insufficient')) {
            code = errors_1.ErrorCode.SWAP_INSUFFICIENT_BALANCE;
        }
        else if (reason.includes('slippage')) {
            code = errors_1.ErrorCode.SWAP_SLIPPAGE_EXCEEDED;
        }
        else if (reason.includes('rejected')) {
            code = errors_1.ErrorCode.SWAP_REJECTED;
        }
        return new exceptions_1.SwapError(message, code, metadata);
    }
    /**
     * Create a Lightning Network error
     */
    static createNodeError(operation, reason, nodeId) {
        const message = nodeId
            ? `Lightning node error for '${nodeId}' during ${operation}: ${reason}`
            : `Lightning node error during ${operation}: ${reason}`;
        const metadata = {
            context: {
                operation,
                reason,
                nodeId
            },
            timestamp: new Date(),
            recoveryActions: [
                'Check Lightning node connectivity',
                'Verify node configuration',
                'Check channel capacity',
                'Contact support if issue persists'
            ]
        };
        let code = errors_1.ErrorCode.LN_NODE_UNREACHABLE;
        if (reason.includes('not configured')) {
            code = errors_1.ErrorCode.LN_NODE_NOT_CONFIGURED;
        }
        else if (reason.includes('channel')) {
            code = errors_1.ErrorCode.LN_CHANNEL_NOT_FOUND;
        }
        else if (reason.includes('capacity')) {
            code = errors_1.ErrorCode.LN_INSUFFICIENT_INBOUND_CAPACITY;
        }
        else if (reason.includes('payment')) {
            code = errors_1.ErrorCode.LN_PAYMENT_FAILED;
        }
        return new exceptions_1.NodeError(message, code, metadata);
    }
    /**
     * Create a timeout error
     */
    static createTimeoutError(operation, timeoutMs, context) {
        const message = `Operation '${operation}' timed out after ${timeoutMs}ms`;
        const metadata = {
            context: {
                operation,
                timeoutMs,
                ...context
            },
            timestamp: new Date(),
            retryable: true,
            retryStrategy: errors_1.RetryStrategy.EXPONENTIAL_BACKOFF,
            recoveryActions: [
                'Try again with a longer timeout',
                'Check network connectivity',
                'Verify server status'
            ]
        };
        return new exceptions_1.TimeoutError(message, errors_1.ErrorCode.OPERATION_TIMEOUT, metadata);
    }
    /**
     * Create a configuration error
     */
    static createConfigError(field, reason, providedValue) {
        const message = `Configuration error for '${field}': ${reason}`;
        const metadata = {
            context: {
                field,
                reason,
                providedValue
            },
            timestamp: new Date(),
            retryable: false,
            recoveryActions: [
                `Check the '${field}' configuration`,
                'Refer to documentation for valid values',
                'Contact support if issue persists'
            ]
        };
        let code = errors_1.ErrorCode.CONFIG_MISSING_REQUIRED_FIELD;
        if (reason.includes('invalid')) {
            code = errors_1.ErrorCode.CONFIG_INVALID_BASE_URL;
        }
        return new exceptions_1.ConfigurationError(message, code, metadata);
    }
    /**
     * Wrap an unknown error into the SDK error system
     */
    static fromUnknownError(error, context, operation) {
        const metadata = {
            context: { operation, errorContext: context },
            timestamp: new Date(),
            recoveryActions: [
                'Try the operation again',
                'Check input parameters',
                'Contact support if issue persists'
            ]
        };
        if (error instanceof errors_1.KaleidoSDKError) {
            return error;
        }
        if (error instanceof Error) {
            const message = context
                ? `${context}: ${error.message}`
                : error.message;
            return new errors_1.KaleidoSDKError(message, errors_1.ErrorCode.UNKNOWN_ERROR, errors_1.ErrorCategory.NETWORK, errors_1.ErrorSeverity.MEDIUM, metadata, error);
        }
        const message = context
            ? `${context}: ${String(error)}`
            : `Unknown error: ${String(error)}`;
        return new errors_1.KaleidoSDKError(message, errors_1.ErrorCode.UNKNOWN_ERROR, errors_1.ErrorCategory.NETWORK, errors_1.ErrorSeverity.MEDIUM, metadata);
    }
}
exports.ErrorFactory = ErrorFactory;
