// @kaleidoswap/kaleidoswap-sdk/dist/types/errors.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.KaleidoSDKError = exports.RetryStrategy = exports.ErrorCategory = exports.ErrorSeverity = exports.ErrorCode = void 0;
/**
 * Error codes for different types of errors that can occur in the SDK
 */
var ErrorCode;
(function (ErrorCode) {
    // Network and connectivity errors (1000-1099)
    ErrorCode["NETWORK_UNREACHABLE"] = "NETWORK_UNREACHABLE";
    ErrorCode["CONNECTION_TIMEOUT"] = "CONNECTION_TIMEOUT";
    ErrorCode["CONNECTION_REFUSED"] = "CONNECTION_REFUSED";
    ErrorCode["DNS_RESOLUTION_FAILED"] = "DNS_RESOLUTION_FAILED";
    ErrorCode["WEBSOCKET_CONNECTION_FAILED"] = "WEBSOCKET_CONNECTION_FAILED";
    ErrorCode["WEBSOCKET_DISCONNECTED"] = "WEBSOCKET_DISCONNECTED";
    // HTTP and API errors (1100-1199)
    ErrorCode["HTTP_BAD_REQUEST"] = "HTTP_BAD_REQUEST";
    ErrorCode["HTTP_UNAUTHORIZED"] = "HTTP_UNAUTHORIZED";
    ErrorCode["HTTP_FORBIDDEN"] = "HTTP_FORBIDDEN";
    ErrorCode["HTTP_NOT_FOUND"] = "HTTP_NOT_FOUND";
    ErrorCode["HTTP_METHOD_NOT_ALLOWED"] = "HTTP_METHOD_NOT_ALLOWED";
    ErrorCode["HTTP_CONFLICT"] = "HTTP_CONFLICT";
    ErrorCode["HTTP_UNPROCESSABLE_ENTITY"] = "HTTP_UNPROCESSABLE_ENTITY";
    ErrorCode["HTTP_TOO_MANY_REQUESTS"] = "HTTP_TOO_MANY_REQUESTS";
    ErrorCode["HTTP_INTERNAL_SERVER_ERROR"] = "HTTP_INTERNAL_SERVER_ERROR";
    ErrorCode["HTTP_BAD_GATEWAY"] = "HTTP_BAD_GATEWAY";
    ErrorCode["HTTP_SERVICE_UNAVAILABLE"] = "HTTP_SERVICE_UNAVAILABLE";
    ErrorCode["HTTP_GATEWAY_TIMEOUT"] = "HTTP_GATEWAY_TIMEOUT";
    // Authentication and authorization errors (1200-1299)
    ErrorCode["AUTH_INVALID_API_KEY"] = "AUTH_INVALID_API_KEY";
    ErrorCode["AUTH_EXPIRED_API_KEY"] = "AUTH_EXPIRED_API_KEY";
    ErrorCode["AUTH_MISSING_API_KEY"] = "AUTH_MISSING_API_KEY";
    ErrorCode["AUTH_INSUFFICIENT_PERMISSIONS"] = "AUTH_INSUFFICIENT_PERMISSIONS";
    // Validation errors (1300-1399)
    ErrorCode["VALIDATION_REQUIRED_FIELD_MISSING"] = "VALIDATION_REQUIRED_FIELD_MISSING";
    ErrorCode["VALIDATION_INVALID_FORMAT"] = "VALIDATION_INVALID_FORMAT";
    ErrorCode["VALIDATION_OUT_OF_RANGE"] = "VALIDATION_OUT_OF_RANGE";
    ErrorCode["VALIDATION_INVALID_ASSET_ID"] = "VALIDATION_INVALID_ASSET_ID";
    ErrorCode["VALIDATION_INVALID_PAIR"] = "VALIDATION_INVALID_PAIR";
    ErrorCode["VALIDATION_INVALID_AMOUNT"] = "VALIDATION_INVALID_AMOUNT";
    // Business logic errors (1400-1499)
    ErrorCode["ASSET_NOT_FOUND"] = "ASSET_NOT_FOUND";
    ErrorCode["ASSET_NOT_SUPPORTED"] = "ASSET_NOT_SUPPORTED";
    ErrorCode["PAIR_NOT_FOUND"] = "PAIR_NOT_FOUND";
    ErrorCode["PAIR_NOT_SUPPORTED"] = "PAIR_NOT_SUPPORTED";
    ErrorCode["INSUFFICIENT_LIQUIDITY"] = "INSUFFICIENT_LIQUIDITY";
    ErrorCode["QUOTE_EXPIRED"] = "QUOTE_EXPIRED";
    ErrorCode["QUOTE_NOT_FOUND"] = "QUOTE_NOT_FOUND";
    ErrorCode["ORDER_NOT_FOUND"] = "ORDER_NOT_FOUND";
    ErrorCode["ORDER_EXPIRED"] = "ORDER_EXPIRED";
    ErrorCode["ORDER_ALREADY_EXECUTED"] = "ORDER_ALREADY_EXECUTED";
    // Swap and trading errors (1500-1599)
    ErrorCode["SWAP_FAILED"] = "SWAP_FAILED";
    ErrorCode["SWAP_TIMEOUT"] = "SWAP_TIMEOUT";
    ErrorCode["SWAP_REJECTED"] = "SWAP_REJECTED";
    ErrorCode["SWAP_INSUFFICIENT_BALANCE"] = "SWAP_INSUFFICIENT_BALANCE";
    ErrorCode["SWAP_PRICE_IMPACT_TOO_HIGH"] = "SWAP_PRICE_IMPACT_TOO_HIGH";
    ErrorCode["SWAP_SLIPPAGE_EXCEEDED"] = "SWAP_SLIPPAGE_EXCEEDED";
    // Lightning Network errors (1600-1699)
    ErrorCode["LN_NODE_UNREACHABLE"] = "LN_NODE_UNREACHABLE";
    ErrorCode["LN_NODE_NOT_CONFIGURED"] = "LN_NODE_NOT_CONFIGURED";
    ErrorCode["LN_CHANNEL_NOT_FOUND"] = "LN_CHANNEL_NOT_FOUND";
    ErrorCode["LN_INSUFFICIENT_INBOUND_CAPACITY"] = "LN_INSUFFICIENT_INBOUND_CAPACITY";
    ErrorCode["LN_INSUFFICIENT_OUTBOUND_CAPACITY"] = "LN_INSUFFICIENT_OUTBOUND_CAPACITY";
    ErrorCode["LN_PAYMENT_FAILED"] = "LN_PAYMENT_FAILED";
    ErrorCode["LN_INVOICE_EXPIRED"] = "LN_INVOICE_EXPIRED";
    // Rate limiting and throttling (1700-1799)
    ErrorCode["RATE_LIMIT_EXCEEDED"] = "RATE_LIMIT_EXCEEDED";
    ErrorCode["QUOTA_EXCEEDED"] = "QUOTA_EXCEEDED";
    ErrorCode["THROTTLED"] = "THROTTLED";
    // Configuration and setup errors (1800-1899)
    ErrorCode["CONFIG_INVALID_BASE_URL"] = "CONFIG_INVALID_BASE_URL";
    ErrorCode["CONFIG_MISSING_REQUIRED_FIELD"] = "CONFIG_MISSING_REQUIRED_FIELD";
    ErrorCode["CONFIG_INVALID_RETRY_CONFIG"] = "CONFIG_INVALID_RETRY_CONFIG";
    // Parsing and serialization errors (1900-1999)
    ErrorCode["PARSE_INVALID_JSON"] = "PARSE_INVALID_JSON";
    ErrorCode["PARSE_INVALID_RESPONSE_FORMAT"] = "PARSE_INVALID_RESPONSE_FORMAT";
    ErrorCode["SERIALIZE_FAILED"] = "SERIALIZE_FAILED";
    // Generic errors (2000-2099)
    ErrorCode["UNKNOWN_ERROR"] = "UNKNOWN_ERROR";
    ErrorCode["OPERATION_TIMEOUT"] = "OPERATION_TIMEOUT";
    ErrorCode["OPERATION_CANCELLED"] = "OPERATION_CANCELLED";
    ErrorCode["NOT_IMPLEMENTED"] = "NOT_IMPLEMENTED";
})(ErrorCode || (exports.ErrorCode = ErrorCode = {}));
/**
 * Error severity levels
 */
var ErrorSeverity;
(function (ErrorSeverity) {
    ErrorSeverity["LOW"] = "low";
    ErrorSeverity["MEDIUM"] = "medium";
    ErrorSeverity["HIGH"] = "high";
    ErrorSeverity["CRITICAL"] = "critical";
})(ErrorSeverity || (exports.ErrorSeverity = ErrorSeverity = {}));
var ErrorCategory;
(function (ErrorCategory) {
    ErrorCategory["NETWORK"] = "network";
    ErrorCategory["HTTP"] = "http";
    ErrorCategory["AUTHENTICATION"] = "authentication";
    ErrorCategory["VALIDATION"] = "validation";
    ErrorCategory["BUSINESS_LOGIC"] = "business_logic";
    ErrorCategory["TRADING"] = "trading";
    ErrorCategory["LIGHTNING_NETWORK"] = "lightning_network";
    ErrorCategory["RATE_LIMITING"] = "rate_limiting";
    ErrorCategory["CONFIGURATION"] = "configuration";
})(ErrorCategory || (exports.ErrorCategory = ErrorCategory = {}));
/**
 * Retry strategy for different types of errors
 */
var RetryStrategy;
(function (RetryStrategy) {
    RetryStrategy["NO_RETRY"] = "no_retry";
    RetryStrategy["IMMEDIATE_RETRY"] = "immediate_retry";
    RetryStrategy["LINEAR_BACKOFF"] = "linear_backoff";
    RetryStrategy["EXPONENTIAL_BACKOFF"] = "exponential_backoff";
    RetryStrategy["CUSTOM"] = "custom";
})(RetryStrategy || (exports.RetryStrategy = RetryStrategy = {}));
/**
 * Base error class for all SDK errors with enhanced metadata
 */
class KaleidoSDKError extends Error {
    constructor(message, code, category, severity = ErrorSeverity.MEDIUM, metadata = {}, cause) {
        super(message);
        this.name = 'KaleidoSDKError';
        this.code = code;
        this.category = category;
        this.severity = severity;
        this.metadata = {
            timestamp: new Date(),
            ...metadata
        };
        this.cause = cause;
        // Ensure proper prototype chain for instanceof checks
        Object.setPrototypeOf(this, KaleidoSDKError.prototype);
    }
    /**
     * Convert error to JSON for logging or debugging
     */
    toJSON() {
        return {
            name: this.name,
            message: this.message,
            code: this.code,
            category: this.category,
            severity: this.severity,
            metadata: this.metadata,
            stack: this.stack,
            cause: this.cause?.message
        };
    }
    /**
     * Get user-friendly error message
     */
    getUserMessage() {
        switch (this.category) {
            case ErrorCategory.NETWORK:
                return 'Network connection issue. Please check your internet connection and try again.';
            case ErrorCategory.AUTHENTICATION:
                return 'Authentication failed. Please check your API key and try again.';
            case ErrorCategory.RATE_LIMITING:
                return 'Too many requests. Please wait a moment and try again.';
            case ErrorCategory.VALIDATION:
                return 'Invalid request data. Please check your input and try again.';
            case ErrorCategory.BUSINESS_LOGIC:
                return 'Operation failed due to business rules. Please check the requirements.';
            case ErrorCategory.TRADING:
                return 'Trading operation failed. Please try again or contact support.';
            default:
                return 'An unexpected error occurred. Please try again later.';
        }
    }
    /**
     * Check if the error is retryable
     */
    isRetryable() {
        return this.metadata.retryable ?? this.getDefaultRetryable();
    }
    /**
     * Get the recommended retry strategy
     */
    getRetryStrategy() {
        return this.metadata.retryStrategy ?? this.getDefaultRetryStrategy();
    }
    /**
     * Get default retryable status based on error category
     */
    getDefaultRetryable() {
        switch (this.category) {
            case ErrorCategory.NETWORK:
            case ErrorCategory.RATE_LIMITING:
                return true;
            case ErrorCategory.HTTP:
                return this.metadata.statusCode ? [500, 502, 503, 504].includes(this.metadata.statusCode) : false;
            case ErrorCategory.AUTHENTICATION:
            case ErrorCategory.VALIDATION:
            case ErrorCategory.BUSINESS_LOGIC:
                return false;
            default:
                return false;
        }
    }
    /**
     * Get default retry strategy based on error category
     */
    getDefaultRetryStrategy() {
        switch (this.category) {
            case ErrorCategory.NETWORK:
                return RetryStrategy.EXPONENTIAL_BACKOFF;
            case ErrorCategory.RATE_LIMITING:
                return RetryStrategy.LINEAR_BACKOFF;
            case ErrorCategory.HTTP:
                return this.isRetryable() ? RetryStrategy.EXPONENTIAL_BACKOFF : RetryStrategy.NO_RETRY;
            default:
                return RetryStrategy.NO_RETRY;
        }
    }
}
exports.KaleidoSDKError = KaleidoSDKError;
