// @kaleidoswap/kaleidoswap-sdk/dist/types/exceptions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ConfigurationError = exports.HttpError = exports.NodeError = exports.QuoteError = exports.PairError = exports.AssetError = exports.WebSocketError = exports.TimeoutError = exports.SwapError = exports.ValidationError = exports.RateLimitError = exports.AuthenticationError = exports.NetworkError = exports.KaleidoError = void 0;
const errors_1 = require("./errors");
// Legacy base error for backward compatibility
class KaleidoError extends Error {
    constructor(message, statusCode, response) {
        super(message);
        this.statusCode = statusCode;
        this.response = response;
        this.name = 'KaleidoError';
    }
}
exports.KaleidoError = KaleidoError;
/**
 * Network-related errors
 */
class NetworkError extends errors_1.KaleidoSDKError {
    constructor(message, code = errors_1.ErrorCode.NETWORK_UNREACHABLE, metadata = {}, cause) {
        super(message, code, errors_1.ErrorCategory.NETWORK, errors_1.ErrorSeverity.HIGH, metadata, cause);
        this.name = 'NetworkError';
    }
}
exports.NetworkError = NetworkError;
/**
 * Authentication and authorization errors
 */
class AuthenticationError extends errors_1.KaleidoSDKError {
    constructor(message, code = errors_1.ErrorCode.AUTH_INVALID_API_KEY, metadata = {}, cause) {
        super(message, code, errors_1.ErrorCategory.AUTHENTICATION, errors_1.ErrorSeverity.HIGH, metadata, cause);
        this.name = 'AuthenticationError';
    }
}
exports.AuthenticationError = AuthenticationError;
/**
 * Rate limiting errors
 */
class RateLimitError extends errors_1.KaleidoSDKError {
    constructor(message, code = errors_1.ErrorCode.RATE_LIMIT_EXCEEDED, metadata = {}, cause) {
        super(message, code, errors_1.ErrorCategory.RATE_LIMITING, errors_1.ErrorSeverity.MEDIUM, {
            retryable: true,
            retryStrategy: 'linear_backoff',
            retryDelay: 1000,
            maxRetries: 3,
            ...metadata
        }, cause);
        this.name = 'RateLimitError';
    }
}
exports.RateLimitError = RateLimitError;
/**
 * Request validation errors
 */
class ValidationError extends errors_1.KaleidoSDKError {
    constructor(message, code = errors_1.ErrorCode.VALIDATION_INVALID_FORMAT, metadata = {}, cause) {
        super(message, code, errors_1.ErrorCategory.VALIDATION, errors_1.ErrorSeverity.MEDIUM, {
            retryable: false,
            ...metadata
        }, cause);
        this.name = 'ValidationError';
    }
}
exports.ValidationError = ValidationError;
/**
 * Trading and swap-related errors
 */
class SwapError extends errors_1.KaleidoSDKError {
    constructor(message, code = errors_1.ErrorCode.SWAP_FAILED, metadata = {}, cause) {
        super(message, code, errors_1.ErrorCategory.TRADING, errors_1.ErrorSeverity.HIGH, metadata, cause);
        this.name = 'SwapError';
    }
}
exports.SwapError = SwapError;
/**
 * Timeout-related errors
 */
class TimeoutError extends errors_1.KaleidoSDKError {
    constructor(message, code = errors_1.ErrorCode.OPERATION_TIMEOUT, metadata = {}, cause) {
        super(message, code, errors_1.ErrorCategory.NETWORK, errors_1.ErrorSeverity.MEDIUM, {
            retryable: true,
            retryStrategy: 'exponential_backoff',
            ...metadata
        }, cause);
        this.name = 'TimeoutError';
    }
}
exports.TimeoutError = TimeoutError;
/**
 * WebSocket-related errors
 */
class WebSocketError extends errors_1.KaleidoSDKError {
    constructor(message, code = errors_1.ErrorCode.WEBSOCKET_CONNECTION_FAILED, metadata = {}, cause) {
        super(message, code, errors_1.ErrorCategory.NETWORK, errors_1.ErrorSeverity.HIGH, {
            retryable: true,
            retryStrategy: 'exponential_backoff',
            ...metadata
        }, cause);
        this.name = 'WebSocketError';
    }
}
exports.WebSocketError = WebSocketError;
/**
 * Asset-related errors
 */
class AssetError extends errors_1.KaleidoSDKError {
    constructor(message, code = errors_1.ErrorCode.ASSET_NOT_FOUND, metadata = {}, cause) {
        super(message, code, errors_1.ErrorCategory.BUSINESS_LOGIC, errors_1.ErrorSeverity.MEDIUM, metadata, cause);
        this.name = 'AssetError';
    }
}
exports.AssetError = AssetError;
/**
 * Trading pair errors
 */
class PairError extends errors_1.KaleidoSDKError {
    constructor(message, code = errors_1.ErrorCode.PAIR_NOT_FOUND, metadata = {}, cause) {
        super(message, code, errors_1.ErrorCategory.BUSINESS_LOGIC, errors_1.ErrorSeverity.MEDIUM, metadata, cause);
        this.name = 'PairError';
    }
}
exports.PairError = PairError;
/**
 * Quote-related errors
 */
class QuoteError extends errors_1.KaleidoSDKError {
    constructor(message, code = errors_1.ErrorCode.QUOTE_EXPIRED, metadata = {}, cause) {
        super(message, code, errors_1.ErrorCategory.BUSINESS_LOGIC, errors_1.ErrorSeverity.MEDIUM, metadata, cause);
        this.name = 'QuoteError';
    }
}
exports.QuoteError = QuoteError;
/**
 * Lightning node errors
 */
class NodeError extends errors_1.KaleidoSDKError {
    constructor(message, code = errors_1.ErrorCode.LN_NODE_UNREACHABLE, metadata = {}, cause) {
        super(message, code, errors_1.ErrorCategory.LIGHTNING_NETWORK, errors_1.ErrorSeverity.HIGH, metadata, cause);
        this.name = 'NodeError';
    }
}
exports.NodeError = NodeError;
/**
 * HTTP-related errors
 */
class HttpError extends errors_1.KaleidoSDKError {
    constructor(message, statusCode, code, metadata = {}, cause) {
        const errorCode = code || HttpError.getErrorCodeFromStatusCode(statusCode);
        const severity = statusCode >= 500 ? errors_1.ErrorSeverity.HIGH : errors_1.ErrorSeverity.MEDIUM;
        super(message, errorCode, errors_1.ErrorCategory.HTTP, severity, {
            statusCode,
            retryable: statusCode >= 500 || statusCode === 429,
            ...metadata
        }, cause);
        this.name = 'HttpError';
    }
    static getErrorCodeFromStatusCode(statusCode) {
        switch (statusCode) {
            case 400: return errors_1.ErrorCode.HTTP_BAD_REQUEST;
            case 401: return errors_1.ErrorCode.HTTP_UNAUTHORIZED;
            case 403: return errors_1.ErrorCode.HTTP_FORBIDDEN;
            case 404: return errors_1.ErrorCode.HTTP_NOT_FOUND;
            case 405: return errors_1.ErrorCode.HTTP_METHOD_NOT_ALLOWED;
            case 409: return errors_1.ErrorCode.HTTP_CONFLICT;
            case 422: return errors_1.ErrorCode.HTTP_UNPROCESSABLE_ENTITY;
            case 429: return errors_1.ErrorCode.HTTP_TOO_MANY_REQUESTS;
            case 500: return errors_1.ErrorCode.HTTP_INTERNAL_SERVER_ERROR;
            case 502: return errors_1.ErrorCode.HTTP_BAD_GATEWAY;
            case 503: return errors_1.ErrorCode.HTTP_SERVICE_UNAVAILABLE;
            case 504: return errors_1.ErrorCode.HTTP_GATEWAY_TIMEOUT;
            default: return errors_1.ErrorCode.UNKNOWN_ERROR;
        }
    }
}
exports.HttpError = HttpError;
/**
 * Configuration-related errors
 */
class ConfigurationError extends errors_1.KaleidoSDKError {
    constructor(message, code = errors_1.ErrorCode.CONFIG_MISSING_REQUIRED_FIELD, metadata = {}, cause) {
        super(message, code, errors_1.ErrorCategory.CONFIGURATION, errors_1.ErrorSeverity.HIGH, {
            retryable: false,
            ...metadata
        }, cause);
        this.name = 'ConfigurationError';
    }
}
exports.ConfigurationError = ConfigurationError;
