// @kaleidoswap/kaleidoswap-sdk/dist/types/orders.js
"use strict";
/**
 * Types for swap order creation and status checking
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.ChannelCheckErrorType = exports.SwapOrderSide = exports.SwapOrderStatus = exports.SwapSettlement = void 0;
var SwapSettlement;
(function (SwapSettlement) {
    SwapSettlement["LIGHTNING"] = "LIGHTNING";
    SwapSettlement["ONCHAIN"] = "ONCHAIN";
})(SwapSettlement || (exports.SwapSettlement = SwapSettlement = {}));
var SwapOrderStatus;
(function (SwapOrderStatus) {
    SwapOrderStatus["OPEN"] = "OPEN";
    SwapOrderStatus["PENDING_PAYMENT"] = "PENDING_PAYMENT";
    SwapOrderStatus["PAID"] = "PAID";
    SwapOrderStatus["EXECUTING"] = "EXECUTING";
    SwapOrderStatus["FILLED"] = "FILLED";
    SwapOrderStatus["CANCELLED"] = "CANCELLED";
    SwapOrderStatus["EXPIRED"] = "EXPIRED";
    SwapOrderStatus["FAILED"] = "FAILED";
})(SwapOrderStatus || (exports.SwapOrderStatus = SwapOrderStatus = {}));
var SwapOrderSide;
(function (SwapOrderSide) {
    SwapOrderSide["BUY"] = "BUY";
    SwapOrderSide["SELL"] = "SELL";
})(SwapOrderSide || (exports.SwapOrderSide = SwapOrderSide = {}));
var ChannelCheckErrorType;
(function (ChannelCheckErrorType) {
    ChannelCheckErrorType["NO_CHANNEL"] = "no_channel";
    ChannelCheckErrorType["INSUFFICIENT_LIQUIDITY"] = "insufficient_liquidity";
    ChannelCheckErrorType["ERROR"] = "error";
})(ChannelCheckErrorType || (exports.ChannelCheckErrorType = ChannelCheckErrorType = {}));
