// @kaleidoswap/kaleidoswap-sdk/dist/client.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.KaleidoClient = void 0;
const client_1 = require("./http/client");
const client_2 = require("./websocket/client");
const ws_1 = __importDefault(require("ws"));
const errorFactory_1 = require("./types/errorFactory");
const index_1 = require("./index");
class KaleidoClient {
    constructor(config) {
        const { nodeUrl, wsUrl, baseUrl, ...apiConfig } = config;
        const finalBaseUrl = baseUrl || process.env.KALEIDO_API_URL || 'https://api.staging.kaleidoswap.com/api/v1';
        this.apiClient = new client_1.HttpClient({ ...apiConfig, baseUrl: finalBaseUrl });
        this.nodeClient = nodeUrl ? new client_1.HttpClient({
            ...apiConfig,
            baseUrl: nodeUrl
        }) : null;
        let wsBaseUrl;
        if (wsUrl) {
            wsBaseUrl = wsUrl;
        }
        else {
            const url = new URL(finalBaseUrl);
            url.protocol = url.protocol === 'https:' ? 'wss:' : 'ws:';
            wsBaseUrl = url.toString();
        }
        this.wsClient = new client_2.WebSocketClient({
            ...apiConfig,
            baseUrl: wsBaseUrl
        });
        if (process.env.DEBUG_WS) {
            console.log('Initialized WebSocket client with URL:', wsBaseUrl);
        }
    }
    ensureNodeClient() {
        if (!this.nodeClient) {
            throw errorFactory_1.ErrorFactory.createNodeError('node operation', 'Node URL is required for this operation. Please provide nodeUrl in the client configuration.');
        }
    }
    async getLspInfo() {
        try {
            return await this.apiClient.get('/lsps1/get_info');
        }
        catch (error) {
            throw errorFactory_1.ErrorFactory.fromUnknownError(error, 'Failed to get LSP info', 'getLspInfo');
        }
    }
    async getLspConnectionUrl() {
        const lspInfo = await this.getLspInfo();
        return lspInfo?.lsp_connection_url;
    }
    async getLspNetworkInfo() {
        try {
            return await this.apiClient.get('/lsps1/network_info');
        }
        catch (error) {
            throw errorFactory_1.ErrorFactory.fromUnknownError(error, 'Failed to get LSP network info', 'getLspNetworkInfo');
        }
    }
    async connectPeer(connectionUrl) {
        this.ensureNodeClient();
        try {
            return await this.nodeClient.post('/connectpeer', { peer_pubkey_and_addr: connectionUrl });
        }
        catch (error) {
            throw errorFactory_1.ErrorFactory.fromUnknownError(error, 'Failed to connect peer', 'connectPeer');
        }
    }
    async getAssetMetadata(assetId) {
        this.ensureNodeClient();
        try {
            return await this.nodeClient.post('/assetmetadata', { asset_id: assetId });
        }
        catch (error) {
            throw new Error(`Failed to get asset metadata: ${error instanceof Error ? error.message : String(error)}`);
        }
    }
    async getNodeInfo() {
        this.ensureNodeClient();
        try {
            return await this.nodeClient.get('/nodeinfo');
        }
        catch (error) {
            throw new index_1.NodeError(`Failed to get node info: ${error instanceof Error ? error.message : String(error)}`);
        }
    }
    async getNodePubkey() {
        const info = await this.getNodeInfo();
        return info.pubkey;
    }
    async assetList() {
        try {
            return await this.apiClient.get('/market/assets');
        }
        catch (error) {
            throw new index_1.AssetError(`Failed to list assets: ${error instanceof Error ? error.message : String(error)}`);
        }
    }
    async pairList() {
        try {
            return await this.apiClient.get('/market/pairs');
        }
        catch (error) {
            throw new index_1.PairError(`Failed to list pairs: ${error instanceof Error ? error.message : String(error)}`);
        }
    }
    async quoteRequest(fromAsset, toAsset, fromAmount, toAmount) {
        try {
            const requestBody = fromAmount ? {
                from_asset: fromAsset,
                to_asset: toAsset,
                from_amount: fromAmount
            }
                : {
                    from_asset: fromAsset,
                    to_asset: toAsset,
                    to_amount: toAmount
                };
            return await this.apiClient.post('/market/quote', requestBody);
        }
        catch (error) {
            throw new index_1.QuoteError(`Failed to get quote: ${error instanceof Error ? error.message : String(error)}`);
        }
    }
    async quoteRequestWS(fromAsset, toAsset, fromAmount, toAmount) {
        if (this.wsClient.getConnectionState() !== ws_1.default.OPEN) {
            await this.wsClient.connect();
        }
        return new Promise((resolve, reject) => {
            let timeoutId;
            const quoteMessage = fromAmount
                ? {
                    action: 'quote_request',
                    from_asset: fromAsset,
                    to_asset: toAsset,
                    from_amount: fromAmount,
                    timestamp: Math.floor(Date.now() / 1000)
                }
                : {
                    action: 'quote_request',
                    from_asset: fromAsset,
                    to_asset: toAsset,
                    to_amount: toAmount,
                    timestamp: Math.floor(Date.now() / 1000)
                };
            console.log('Sending quote message:', quoteMessage);
            timeoutId = setTimeout(() => {
                this.wsClient.off('quote_response', quoteHandler);
                reject(new index_1.TimeoutError('Quote request timed out'));
            }, 30000);
            const quoteHandler = (response) => {
                console.log('quoteHandler received:', response);
                clearTimeout(timeoutId);
                this.wsClient.off('quote_response', quoteHandler);
                if (response.error) {
                    reject(new index_1.QuoteError(response.error.message || 'Failed to get quote'));
                }
                else if (response.data) {
                    resolve(response.data);
                }
                else {
                    resolve(response);
                }
            };
            this.wsClient.on('quote_response', quoteHandler);
            this.wsClient.send(quoteMessage).catch((error) => {
                clearTimeout(timeoutId);
                this.wsClient.off('quote_response', quoteHandler);
                reject(new index_1.WebSocketError(`Failed to send quote request: ${error.message}`));
            });
        });
    }
    async initMakerSwap(request) {
        try {
            return await this.apiClient.post('/swaps/init', {
                rfq_id: request.rfq_id,
                from_asset: request.from_asset,
                to_asset: request.to_asset,
                from_amount: request.from_amount,
                to_amount: request.to_amount,
            });
        }
        catch (error) {
            throw new index_1.SwapError(`Failed to initialize maker swap: ${error instanceof Error ? error.message : String(error)}`);
        }
    }
    async executeMakerSwap(request) {
        try {
            console.log('Executing maker swap with request:', JSON.stringify({
                swapstring: request.swapstring ? `${request.swapstring.substring(0, 20)}...` : 'undefined',
                payment_hash: request.payment_hash,
                taker_pubkey: request.taker_pubkey
            }, null, 2));
            const response = await this.apiClient.post('/swaps/execute', {
                swapstring: request.swapstring,
                payment_hash: request.payment_hash,
                taker_pubkey: request.taker_pubkey
            });
            console.log('Maker swap executed successfully:', response);
            return response;
        }
        catch (error) {
            console.error('Error executing maker swap:', error);
            throw new index_1.SwapError(`Failed to execute maker swap: ${error instanceof Error ? error.message : String(error)}`);
        }
    }
    async whitelistTrade(swapstring) {
        this.ensureNodeClient();
        try {
            return await this.nodeClient.post('/taker', {
                swapstring,
            });
        }
        catch (error) {
            throw new index_1.SwapError(`Failed to whitelist trade: ${error instanceof Error ? error.message : String(error)}`);
        }
    }
    async atomicSwapStatus(orderId) {
        this.ensureNodeClient();
        try {
            return await this.nodeClient.post(`/swaps/atomic/status`, {
                order_id: orderId
            });
        }
        catch (error) {
            throw new index_1.SwapError(`Failed to get swap status: ${error instanceof Error ? error.message : String(error)}`);
        }
    }
    async waitForSwapCompletion(paymentHash, timeoutSeconds = 300, pollIntervalSeconds = 5) {
        const startTime = Date.now();
        const timeoutMs = timeoutSeconds * 1000;
        const pollIntervalMs = pollIntervalSeconds * 1000;
        while (true) {
            const swap = await this.atomicSwapStatus(paymentHash);
            // Check if swap has reached a terminal state
            if (swap.status && ['Succeeded', 'Failed', 'Expired'].includes(swap.status)) {
                return swap;
            }
            // Check if we've exceeded the timeout
            if (Date.now() - startTime >= timeoutMs) {
                throw new index_1.TimeoutError(`Swap ${paymentHash} did not complete within ${timeoutSeconds} seconds`);
            }
            // Wait before polling again
            await new Promise(resolve => setTimeout(resolve, pollIntervalMs));
        }
    }
    async createOrder(request) {
        try {
            return await this.apiClient.post('/swaps/orders', request);
        }
        catch (error) {
            throw new index_1.SwapError(`Failed to create swap order: ${error instanceof Error ? error.message : String(error)}`);
        }
    }
    async swapOrderStatus(request) {
        try {
            return await this.apiClient.post('/swaps/orders/status', request);
        }
        catch (error) {
            throw new index_1.SwapError(`Failed to get swap order status: ${error instanceof Error ? error.message : String(error)}`);
        }
    }
    async swapOrderAnalytic() {
        try {
            return await this.apiClient.get('/swaps/orders/analytics');
        }
        catch (error) {
            throw new index_1.SwapError(`Failed to get swap orders analytics: ${error instanceof Error ? error.message : String(error)}`);
        }
    }
}
exports.KaleidoClient = KaleidoClient;
