// @kaleidoswap/kaleidoswap-sdk/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OrderSettlement = exports.createPrecisionHandler = exports.PrecisionHandler = exports.createAssetPairMapper = exports.AssetPairMapper = exports.createRetryWrapper = exports.withRetry = exports.retry = exports.HttpClient = exports.WebSocketClient = exports.ConfigurationError = exports.HttpError = exports.NodeError = exports.QuoteError = exports.PairError = exports.AssetError = exports.WebSocketError = exports.TimeoutError = exports.SwapError = exports.ValidationError = exports.RateLimitError = exports.AuthenticationError = exports.NetworkError = exports.KaleidoError = exports.ErrorFactory = exports.KaleidoSDKError = exports.RetryStrategy = exports.ErrorCategory = exports.ErrorSeverity = exports.ErrorCode = exports.KaleidoClient = void 0;
// Client
var client_1 = require("./client");
Object.defineProperty(exports, "KaleidoClient", { enumerable: true, get: function () { return client_1.KaleidoClient; } });
// Enhanced error handling system
var errors_1 = require("./types/errors");
// Error types and enums
Object.defineProperty(exports, "ErrorCode", { enumerable: true, get: function () { return errors_1.ErrorCode; } });
Object.defineProperty(exports, "ErrorSeverity", { enumerable: true, get: function () { return errors_1.ErrorSeverity; } });
Object.defineProperty(exports, "ErrorCategory", { enumerable: true, get: function () { return errors_1.ErrorCategory; } });
Object.defineProperty(exports, "RetryStrategy", { enumerable: true, get: function () { return errors_1.RetryStrategy; } });
// Base error class
Object.defineProperty(exports, "KaleidoSDKError", { enumerable: true, get: function () { return errors_1.KaleidoSDKError; } });
// Error factory
var errorFactory_1 = require("./types/errorFactory");
Object.defineProperty(exports, "ErrorFactory", { enumerable: true, get: function () { return errorFactory_1.ErrorFactory; } });
// Legacy and specific error classes
var exceptions_1 = require("./types/exceptions");
// Legacy base error (for backward compatibility)
Object.defineProperty(exports, "KaleidoError", { enumerable: true, get: function () { return exceptions_1.KaleidoError; } });
// Specific error types
Object.defineProperty(exports, "NetworkError", { enumerable: true, get: function () { return exceptions_1.NetworkError; } });
Object.defineProperty(exports, "AuthenticationError", { enumerable: true, get: function () { return exceptions_1.AuthenticationError; } });
Object.defineProperty(exports, "RateLimitError", { enumerable: true, get: function () { return exceptions_1.RateLimitError; } });
Object.defineProperty(exports, "ValidationError", { enumerable: true, get: function () { return exceptions_1.ValidationError; } });
Object.defineProperty(exports, "SwapError", { enumerable: true, get: function () { return exceptions_1.SwapError; } });
Object.defineProperty(exports, "TimeoutError", { enumerable: true, get: function () { return exceptions_1.TimeoutError; } });
Object.defineProperty(exports, "WebSocketError", { enumerable: true, get: function () { return exceptions_1.WebSocketError; } });
Object.defineProperty(exports, "AssetError", { enumerable: true, get: function () { return exceptions_1.AssetError; } });
Object.defineProperty(exports, "PairError", { enumerable: true, get: function () { return exceptions_1.PairError; } });
Object.defineProperty(exports, "QuoteError", { enumerable: true, get: function () { return exceptions_1.QuoteError; } });
Object.defineProperty(exports, "NodeError", { enumerable: true, get: function () { return exceptions_1.NodeError; } });
Object.defineProperty(exports, "HttpError", { enumerable: true, get: function () { return exceptions_1.HttpError; } });
Object.defineProperty(exports, "ConfigurationError", { enumerable: true, get: function () { return exceptions_1.ConfigurationError; } });
// WebSocket
var client_2 = require("./websocket/client");
Object.defineProperty(exports, "WebSocketClient", { enumerable: true, get: function () { return client_2.WebSocketClient; } });
// HTTP
var client_3 = require("./http/client");
Object.defineProperty(exports, "HttpClient", { enumerable: true, get: function () { return client_3.HttpClient; } });
// Utils
var retry_1 = require("./utils/retry");
Object.defineProperty(exports, "retry", { enumerable: true, get: function () { return retry_1.retry; } });
Object.defineProperty(exports, "withRetry", { enumerable: true, get: function () { return retry_1.withRetry; } });
Object.defineProperty(exports, "createRetryWrapper", { enumerable: true, get: function () { return retry_1.createRetryWrapper; } });
var assetPairMapper_1 = require("./utils/assetPairMapper");
Object.defineProperty(exports, "AssetPairMapper", { enumerable: true, get: function () { return assetPairMapper_1.AssetPairMapper; } });
Object.defineProperty(exports, "createAssetPairMapper", { enumerable: true, get: function () { return assetPairMapper_1.createAssetPairMapper; } });
var precisionHandler_1 = require("./utils/precisionHandler");
Object.defineProperty(exports, "PrecisionHandler", { enumerable: true, get: function () { return precisionHandler_1.PrecisionHandler; } });
Object.defineProperty(exports, "createPrecisionHandler", { enumerable: true, get: function () { return precisionHandler_1.createPrecisionHandler; } });
// Orders
var OrderSettlement;
(function (OrderSettlement) {
    OrderSettlement["LIGHTNING"] = "LIGHTNING";
    OrderSettlement["ONCHAIN"] = "ONCHAIN";
})(OrderSettlement || (exports.OrderSettlement = OrderSettlement = {}));
