// @kaleidoswap/kaleidoswap-sdk/dist/testUtils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.withWebSocketTest = exports.testWhiteListTrade = exports.assetListResponse = exports.getPairAssetIds = exports.createTestClient = exports.testConfig = void 0;
const client_1 = require("./client");
exports.testConfig = {
    apiKey: process.env.TEST_API_KEY || '',
    get wsUrl() {
        // Use the same default as client.ts
        const baseUrl = this.baseUrl || process.env.KALEIDO_API_URL || 'https://api.staging.kaleidoswap.com/api/v1';
        const url = new URL(baseUrl);
        url.protocol = url.protocol === 'https:' ? 'wss:' : 'ws:';
        url.pathname = '/api/v1/market/ws/testclient';
        return url.toString();
    }
};
const createTestClient = () => {
    const config = exports.testConfig;
    // Build the client config, only including baseUrl if it's defined
    const clientConfig = {
        apiKey: config.apiKey,
        wsUrl: config.wsUrl
    };
    // Only add baseUrl if it's explicitly set
    if (config.baseUrl) {
        clientConfig.baseUrl = config.baseUrl;
    }
    const client = new client_1.KaleidoClient(clientConfig);
    if (process.env.DEBUG_WS) {
        console.log('Created test client with WebSocket URL:', config.wsUrl);
    }
    return client;
};
exports.createTestClient = createTestClient;
const getPairAssetIds = async (client) => {
    const response = await client.pairList();
    const pairs = response.pairs;
    if (pairs && pairs.length > 0) {
        const pair = pairs[0];
        return {
            baseAssetId: pair.base_asset_id,
            quoteAssetId: pair.quote_asset_id
        };
    }
    throw new Error("No pair was found.");
};
exports.getPairAssetIds = getPairAssetIds;
const assetListResponse = async (client) => {
    try {
        const assetsResponse = await client.assetList();
        const assetsList = assetsResponse.assets || [];
        // Find USDT asset_id
        const toAssetObj = assetsList.find(a => a.ticker === "USDT");
        const toAsset = toAssetObj?.asset_id;
        // Hardcode BTC as fromAsset (by ticker, not asset_id)
        const fromAsset = "BTC";
        const fromAmount = 10000000; // 0.1 BTC in satoshis
        if (!toAsset) {
            throw new Error('Required asset (USDT) for testing not found');
        }
        return { fromAsset, toAsset, fromAmount };
    }
    catch (error) {
        console.error('Error in assetListResponse:', error);
        throw error;
    }
};
exports.assetListResponse = assetListResponse;
const testWhiteListTrade = async (client, maxRetries = 3, retryDelay = 1000) => {
    for (let attempt = 1; attempt <= maxRetries; attempt++) {
        try {
            console.log(`Attempt ${attempt} to initialize swap...`);
            // Get assets and quote
            const assetsResponse = await (0, exports.assetListResponse)(client);
            const quote = await client.quoteRequestWS(assetsResponse.fromAsset, assetsResponse.toAsset, assetsResponse.fromAmount);
            console.log('Got quote, initializing swap...');
            // Initialize the swap
            const initResult = await client.initMakerSwap({
                rfq_id: quote.rfq_id,
                from_asset: quote.from_asset,
                from_amount: quote.from_amount,
                to_asset: quote.to_asset,
                to_amount: quote.to_amount
            }); // the response type only has SwapResponse...
            if (!initResult.swapstring) {
                throw new Error('No swapstring returned from initMakerSwap');
            }
            console.log('Swap initialized, whitelisting trade...');
            // Whitelist the trade
            await client.whitelistTrade(initResult.swapstring);
            return initResult;
        }
        catch (error) {
            const errorMessage = error instanceof Error ? error.message : String(error);
            // Check if we should retry
            const isRetryable = errorMessage.includes('Not enough liquidity') ||
                errorMessage.includes('400') ||
                errorMessage.includes('Bad Request');
            if (!isRetryable || attempt >= maxRetries) {
                console.error(`Failed to initialize swap (attempt ${attempt}/${maxRetries}):`, errorMessage);
                throw error;
            }
            console.warn(`Retryable error (${attempt}/${maxRetries}): ${errorMessage}. Retrying in ${retryDelay}ms...`);
            await new Promise(resolve => setTimeout(resolve, retryDelay));
        }
    }
    return null;
};
exports.testWhiteListTrade = testWhiteListTrade;
const withWebSocketTest = async (testFn, timeout = 30000) => {
    const client = (0, exports.createTestClient)();
    try {
        await new Promise((resolve, reject) => {
            const timer = setTimeout(() => {
                reject(new Error('WebSocket test timed out'));
            }, timeout);
            testFn(client)
                .then(() => {
                clearTimeout(timer);
                resolve();
            })
                .catch(reject);
        });
    }
    finally {
        // Clean up WebSocket connection
        if (client.wsClient) {
            await client.wsClient.disconnect();
        }
    }
};
exports.withWebSocketTest = withWebSocketTest;
