// @kaleidoswap/kaleidoswap-sdk/dist/websocket/client.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.WebSocketClient = void 0;
const errorFactory_1 = require("../types/errorFactory");
const ws_1 = __importDefault(require("ws"));
// Debug logging helper
const debug = (...args) => {
    if (process.env.DEBUG_WS) {
        console.log('[WebSocket]', ...args);
    }
};
/**
 * WebSocket client for real-time updates from the KaleidoSwap API
 */
class WebSocketClient {
    /**
     * Checks if the WebSocket is connected
     * @returns boolean indicating if the WebSocket is connected
     */
    isConnected() {
        return this.ws?.readyState === ws_1.default.OPEN;
    }
    /**
     * Creates a new WebSocket client instance
     * @param config - Configuration for the client
     */
    constructor(config) {
        this.ws = null;
        this.pingTimer = null;
        this.reconnectTimer = null;
        this.messageHandlers = new Map();
        this.reconnectAttempts = 0;
        this.isConnecting = false;
        this.baseUrl = config.baseUrl.replace(/^http/, 'ws');
        this.pingInterval = config.pingInterval || 30000; // 30 seconds
        this.pingTimeout = config.pingTimeout || 10000; // 10 seconds
        this.reconnectInterval = config.reconnectInterval || 5000; // 5 seconds
        this.maxReconnectAttempts = config.maxReconnectAttempts || 5;
        // Set up headers
        this.headers = {
            'Content-Type': 'application/json',
            'Accept': 'application/json'
        };
        // Add API key if provided
        if (config.apiKey) {
            this.headers['Authorization'] = `Bearer ${config.apiKey}`;
        }
        if (process.env.DEBUG_WS) {
            debug(`Initializing WebSocket client for ${this.baseUrl}`);
            if (config.apiKey) {
                debug('Using API key authentication');
            }
        }
    }
    /**
     * Gets the headers for WebSocket connection
     * @returns Object containing connection headers
     */
    //private getHeaders(): Record<string, string> {
    //  const headers: Record<string, string> = {};
    //  if (this.apiKey) {
    //    headers['Authorization'] = `Bearer ${this.apiKey}`;
    //  }
    //  return headers;
    //}
    logConnectionState(event) {
        if (process.env.DEBUG_WS) {
            const state = this.ws?.readyState;
            let stateStr = 'UNKNOWN';
            switch (state) {
                case ws_1.default.CONNECTING:
                    stateStr = 'CONNECTING';
                    break;
                case ws_1.default.OPEN:
                    stateStr = 'OPEN';
                    break;
                case ws_1.default.CLOSING:
                    stateStr = 'CLOSING';
                    break;
                case ws_1.default.CLOSED:
                    stateStr = 'CLOSED';
                    break;
            }
            debug(`${event} - State: ${stateStr}`);
        }
    }
    /**
     * Establishes a WebSocket connection
     * @throws {WebSocketError} If connection fails
     */
    async connect() {
        if (this.isConnected() || this.isConnecting) {
            debug('Connection already in progress or established');
            return;
        }
        this.isConnecting = true;
        // For ws library, we need to handle headers differently
        const wsOptions = {};
        // Add headers if needed (some WebSocket servers support this)
        if (Object.keys(this.headers).length > 0) {
            wsOptions.headers = this.headers;
        }
        // If using an API key, add it to the URL as a query parameter
        // This is a common pattern when headers aren't supported
        let wsUrl = this.baseUrl;
        if (this.headers['Authorization']) {
            const url = new URL(wsUrl);
            url.searchParams.append('token', this.headers['Authorization'].replace('Bearer ', ''));
            wsUrl = url.toString();
        }
        debug(`Connecting to WebSocket: ${wsUrl}`);
        if (process.env.DEBUG_WS) {
            debug('Using options:', JSON.stringify(wsOptions, null, 2));
        }
        try {
            this.ws = new ws_1.default(wsUrl, wsOptions);
            this.logConnectionState('After WebSocket creation');
            await new Promise((resolve, reject) => {
                if (!this.ws) {
                    const error = errorFactory_1.ErrorFactory.fromWebSocketError(new Error('WebSocket instance is null'), 'connection initialization');
                    debug(error.message);
                    reject(error);
                    return;
                }
                const timeout = setTimeout(() => {
                    const error = errorFactory_1.ErrorFactory.createTimeoutError('WebSocket connection', this.pingTimeout, { wsUrl });
                    debug(error.message);
                    reject(error);
                }, this.pingTimeout);
                this.ws.onopen = () => {
                    clearTimeout(timeout);
                    this.reconnectAttempts = 0;
                    this.logConnectionState('onopen');
                    this.startPing();
                    resolve();
                };
                this.ws.onerror = (event) => {
                    clearTimeout(timeout);
                    const errorMessage = event.message || 'Unknown WebSocket error';
                    const error = errorFactory_1.ErrorFactory.fromWebSocketError(new Error(errorMessage), 'connection error');
                    debug('WebSocket error:', error);
                    reject(error);
                };
                this.ws.onmessage = this.handleMessage.bind(this);
                this.ws.onclose = this.handleClose.bind(this);
            });
        }
        catch (error) {
            this.isConnecting = false;
            debug('Connection failed:', error);
            throw error;
        }
        finally {
            if (!this.isConnected()) {
                this.isConnecting = false;
            }
        }
    }
    /**
     * Starts sending periodic ping messages
     */
    startPing() {
        this.pingTimer = setInterval(() => {
            if (this.ws?.readyState === ws_1.default.OPEN) {
                this.ws.ping();
            }
        }, this.pingInterval);
    }
    stopPing() {
        if (this.pingTimer) {
            clearInterval(this.pingTimer);
            this.pingTimer = null;
        }
    }
    /**
     * Handles incoming WebSocket messages
     * @param event - Message event
     */
    handleMessage(event) {
        try {
            const data = event.data.toString();
            if (process.env.DEBUG_WS) {
                debug('Received message:', data);
            }
            const message = JSON.parse(data);
            this.emit(message.action, message.data);
        }
        catch (error) {
            const errorMessage = error instanceof Error ? error.message : String(error);
            debug('Error handling message:', errorMessage);
            this.emit('error', { action: 'message_parse_error', error: errorMessage });
        }
    }
    /**
     * Handles WebSocket connection close
     */
    handleClose(event) {
        this.logConnectionState('onclose');
        debug(`Connection closed: ${event.code} ${event.reason || 'No reason provided'}`);
        this.stopPing();
        // Only attempt reconnect if this was an unexpected closure
        if (event.code !== 1000) { // 1000 is normal closure
            this.attemptReconnect();
        }
        else {
            debug('Normal WebSocket closure, not reconnecting');
        }
    }
    attemptReconnect() {
        if (this.reconnectAttempts >= this.maxReconnectAttempts) {
            const error = errorFactory_1.ErrorFactory.fromWebSocketError(new Error(`Max reconnection attempts (${this.maxReconnectAttempts}) reached`), 'reconnection failure');
            debug(error.message);
            this.emit('error', { action: 'max_reconnect_attempts', error: error.message });
            return;
        }
        const delay = Math.min(this.reconnectInterval * Math.pow(2, this.reconnectAttempts - 1), 30000 // Max 30 seconds between retries
        );
        debug(`Attempting to reconnect (${this.reconnectAttempts + 1}/${this.maxReconnectAttempts}) in ${delay}ms`);
        this.reconnectTimer = setTimeout(() => {
            this.reconnectAttempts++;
            this.connect().catch(error => {
                debug('Reconnection attempt failed:', error);
            });
        }, delay);
    }
    /**
     * Closes the WebSocket connection
     */
    async disconnect(code = 1000, reason) {
        debug(`Disconnecting${reason ? ` (${reason})` : ''}`);
        this.stopPing();
        if (this.reconnectTimer) {
            clearTimeout(this.reconnectTimer);
            this.reconnectTimer = null;
        }
        if (this.ws) {
            return new Promise((resolve) => {
                if (this.ws) {
                    this.ws.once('close', () => {
                        debug('Disconnected successfully');
                        this.ws = null;
                        resolve();
                    });
                    if (this.ws.readyState === ws_1.default.OPEN) {
                        this.ws.close(code, reason);
                    }
                    else {
                        this.ws.terminate();
                        this.ws = null;
                        resolve();
                    }
                }
                else {
                    resolve();
                }
            });
        }
        return Promise.resolve();
    }
    /**
     * Sends a message over the WebSocket connection
     * @param message - Message to send
     * @throws {WebSocketError} If not connected
     */
    async send(message) {
        if (!this.isConnected()) {
            debug('Not connected, attempting to connect...');
            await this.connect();
        }
        if (!this.ws) {
            const error = errorFactory_1.ErrorFactory.fromWebSocketError(new Error('WebSocket is not connected'), 'send message');
            debug('Send failed:', error);
            throw error;
        }
        try {
            const messageStr = JSON.stringify(message);
            if (process.env.DEBUG_WS) {
                debug('Sending message:', messageStr);
            }
            this.ws.send(messageStr);
        }
        catch (error) {
            const errorMessage = error instanceof Error ? error.message : String(error);
            debug('Error sending message:', errorMessage);
            throw errorFactory_1.ErrorFactory.fromWebSocketError(error instanceof Error ? error : new Error(errorMessage), 'send message');
        }
    }
    /**
     * Registers a handler for a specific message action
     * @param action - Message action to handle
     * @param handler - Handler function
     */
    on(event, handler) {
        if (!this.messageHandlers.has(event)) {
            this.messageHandlers.set(event, new Set());
        }
        const handlers = this.messageHandlers.get(event);
        handlers.add(handler);
        // Return unsubscribe function
        return () => {
            handlers.delete(handler);
            if (handlers.size === 0) {
                this.messageHandlers.delete(event);
            }
        };
    }
    /**
     * Removes a handler for a specific message action
     * @param action - Message action
     * @param handler - Handler function to remove
     */
    off(event, handler) {
        const handlers = this.messageHandlers.get(event);
        if (handlers) {
            handlers.delete(handler);
            if (handlers.size === 0) {
                this.messageHandlers.delete(event);
            }
        }
    }
    emit(event, data) {
        const handlers = this.messageHandlers.get(event);
        if (handlers) {
            handlers.forEach(handler => handler(data));
        }
    }
    /**
     * Subscribes to updates for a trading pair
     * @param pairId - Trading pair identifier
     * @throws {WebSocketError} If not connected
     */
    async subscribe(pairId) {
        await this.send({
            action: 'subscribe',
            data: { pairId }
        });
    }
    /**
     * Unsubscribes from updates for a trading pair
     * @param pairId - Trading pair identifier
     * @throws {WebSocketError} If not connected
     */
    async unsubscribe(pairId) {
        await this.send({
            action: 'unsubscribe',
            data: { pairId }
        });
    }
    /**
     * Gets the current WebSocket connection state
     * @returns The current WebSocket ready state or undefined if not connected
     */
    getConnectionState() {
        return this.ws?.readyState;
    }
}
exports.WebSocketClient = WebSocketClient;
