// @kaleidoswap/kaleidoswap-sdk/dist/__tests__/client.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
//import { WebSocket } from 'ws';
const testUtils_1 = require("../testUtils");
describe('KaleidoClient', () => {
    const client = (0, testUtils_1.createTestClient)();
    describe('node operations', () => {
        it('should get node info', async () => {
            const info = await client.getNodeInfo();
            expect(info).toHaveProperty('pubkey');
            expect(typeof info.pubkey).toBe('string');
            //console.log('Node info:', info);
        });
        it('should get node pubkey', async () => {
            const pubkey = await client.getNodePubkey();
            expect(typeof pubkey).toBe('string');
            expect(pubkey.length).toBeGreaterThan(0);
            //console.log('Node pubkey:', pubkey);
        });
        it('should connect to peer', async () => {
            try {
                const lspConnectionUrl = await client.getLspConnectionUrl();
                if (lspConnectionUrl && typeof lspConnectionUrl === 'string') {
                    const result = await client.connectPeer(lspConnectionUrl);
                    expect(result).toBeDefined();
                }
                else {
                    console.warn('Test skipped: No LSP connection URL available');
                }
            }
            catch (error) {
                console.warn('Test skipped: Could not connect to peer:', error);
            }
        });
        // TODO: test for error handling -- not working
        //it('should handle node errors', async () => {
        //  jest.spyOn(client['nodeClient'], 'get').mockRejectedValue(new Error('Node error'));
        //  await expect(client.getNodeInfo()).rejects.toThrow(NodeError);
        //});
    });
    describe('LSP operations', () => {
        it('should get LSP info', async () => {
            const lspInfo = await client.getLspInfo();
            expect(lspInfo).toBeDefined();
            expect(lspInfo).toHaveProperty('lsp_connection_url');
            expect(typeof lspInfo.lsp_connection_url).toBe('string');
        });
        it('should get LSP connection URL', async () => {
            const connectionUrl = await client.getLspConnectionUrl();
            expect(connectionUrl).toBeDefined();
            expect(typeof connectionUrl).toBe('string');
            expect(connectionUrl.length).toBeGreaterThan(0);
        });
        it('should get LSP network info', async () => {
            const networkInfo = await client.getLspNetworkInfo();
            expect(networkInfo).toBeDefined();
            expect(networkInfo).toHaveProperty('network');
        });
    });
    describe('asset operations', () => {
        it('should get list assets', async () => {
            const result = await client.assetList();
            expect(result).toHaveProperty('assets');
            expect(result).toHaveProperty('network');
            expect(result).toHaveProperty('timestamp');
        });
        it('should get asset metadata', async () => {
            try {
                const assets = await client.assetList();
                if (assets.assets && assets.assets.length > 0 && assets.assets[0].asset_id) {
                    const assetId = assets.assets[0].asset_id;
                    const metadata = await client.getAssetMetadata(assetId);
                    expect(metadata).toBeDefined();
                    expect(metadata).toHaveProperty('name');
                    expect(typeof metadata.name).toBe('string');
                }
                else {
                    console.warn('Test skipped: No assets available');
                }
            }
            catch (error) {
                console.warn('Test skipped: Could not get asset metadata:', error);
            }
        });
    });
    describe('trading pair operations', () => {
        it('should list pairs', async () => {
            const result = await client.pairList();
            expect(result).toHaveProperty('pairs');
        });
    });
    describe('quote operations', () => {
        it('should get quote', async () => {
            try {
                const assetsResponse = await (0, testUtils_1.assetListResponse)(client);
                const result = await client.quoteRequest(assetsResponse.fromAsset, assetsResponse.toAsset, assetsResponse.fromAmount);
                expect(result).toBeDefined();
                expect(result).toHaveProperty('rfq_id');
                expect(result).toHaveProperty('to_amount');
            }
            catch {
                const error = new Error('Skipping test: Unable to get quote, possibly due to insufficient liquidity or misconfiguration.');
                console.warn(error.message);
            }
        });
    });
    describe('WebSocket operations', () => {
        it('should get quote via websocket', async () => {
            try {
                const assetsResponse = await (0, testUtils_1.assetListResponse)(client);
                const result = await client.quoteRequestWS(assetsResponse.fromAsset, assetsResponse.toAsset, assetsResponse.fromAmount);
                expect(result).toBeDefined();
                expect(result).toHaveProperty('rfq_id');
                expect(result).toHaveProperty('to_amount');
            }
            catch {
                // TODO: catch error
            }
        });
        // TODO: test for error handling -- not working
    });
    describe('swap operations', () => {
        it('should initialize maker swap', async () => {
            const assetsResponse = await (0, testUtils_1.assetListResponse)(client);
            const quote = await client.quoteRequestWS(assetsResponse.fromAsset, assetsResponse.toAsset, 100000000);
            const init_result = await client.initMakerSwap({
                rfq_id: quote.rfq_id,
                from_asset: quote.from_asset,
                to_asset: quote.to_asset,
                from_amount: quote.from_amount,
                to_amount: quote.to_amount
            });
            expect(init_result).toBeDefined();
            expect(init_result).toHaveProperty('payment_hash');
            expect(init_result).toHaveProperty('swapstring');
        });
        it('should whitelist a trade', async () => {
            const assetsResponse = await (0, testUtils_1.assetListResponse)(client);
            const quote = await client.quoteRequestWS(assetsResponse.fromAsset, assetsResponse.toAsset, 100000000);
            const initResult = await client.initMakerSwap({
                rfq_id: quote.rfq_id,
                from_asset: quote.from_asset,
                to_asset: quote.to_asset,
                from_amount: quote.from_amount,
                to_amount: quote.to_amount
            });
            expect(initResult).toBeDefined();
            expect(initResult).toHaveProperty('payment_hash');
            expect(initResult).toHaveProperty('swapstring');
            if (!initResult.payment_hash || !initResult.swapstring) {
                throw new Error('Missing `payment_hash` and `swapstring` in initMakerSwap response');
            }
            const result = await client.whitelistTrade(initResult.swapstring);
            expect(result).toBeDefined();
            expect(result).toEqual({});
        });
        it('should confirm complete swap', async () => {
            try {
                console.log('Starting swap test...');
                // Initialize and whitelist the trade with more retries
                console.log('Initializing and whitelisting trade...');
                const initResult = await (0, testUtils_1.testWhiteListTrade)(client, 5, 2000); // 5 retries, 2s delay
                if (!initResult) {
                    console.warn('Test skipped: Failed to initialize and whitelist trade after multiple attempts');
                    return; // Skip the test if we can't initialize the trade
                }
                if (!initResult.swapstring || !initResult.payment_hash) {
                    console.warn('Test skipped: Missing required swap data');
                    return;
                }
                console.log('Trade initialized successfully:', {
                    hasSwapstring: !!initResult.swapstring,
                    hasPaymentHash: !!initResult.payment_hash
                });
                try {
                    // Get taker's pubkey
                    console.log('Getting taker pubkey...');
                    const takerPubkey = await client.getNodePubkey();
                    console.log('Taker pubkey:', takerPubkey);
                    // Execute the swap
                    console.log('Executing maker swap...');
                    const executeResult = await client.executeMakerSwap({
                        swapstring: initResult.swapstring,
                        payment_hash: initResult.payment_hash,
                        taker_pubkey: takerPubkey
                    });
                    console.log('Swap execution result:', executeResult);
                    expect(executeResult).toHaveProperty('assets');
                }
                catch (swapError) {
                }
            }
            catch (error) {
                console.error('Swap test failed:', error);
                // Handle specific error cases
                const errorMessage = error instanceof Error ? error.message : String(error);
                if (errorMessage.includes('Not enough liquidity')) {
                    console.warn('Test skipped: Not enough liquidity for swap');
                    return; // Skip the test instead of failing
                }
                if (errorMessage.includes('400') || errorMessage.includes('Bad Request')) {
                    console.warn('Test skipped: Bad request - likely due to liquidity or other server-side validation');
                    return;
                }
                throw error;
            }
        }, 30000);
    });
});
