// @kaleidoswap/kaleidoswap-sdk/dist/utils/assetPairMapper.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AssetPairMapper = void 0;
exports.createAssetPairMapper = createAssetPairMapper;
/**
 * Asset pair mapping utility
 * Creates a comprehensive asset list from trading pairs data
 */
class AssetPairMapper {
    constructor(pairsResponse) {
        this.assetMap = new Map();
        this.pairs = pairsResponse.pairs;
        this.buildAssetMap();
    }
    /**
     * Build asset map from pairs data
     */
    buildAssetMap() {
        this.pairs.forEach(pair => {
            if (!pair.is_active)
                return;
            // Process base asset
            this.processAsset({
                asset_id: pair.base_asset_id,
                ticker: pair.base_asset,
                name: pair.base_asset,
                precision: pair.base_precision,
                is_active: pair.is_active,
                min_order_size: pair.min_base_order_size,
                max_order_size: pair.max_base_order_size,
                trading_partner: pair.quote_asset_id
            });
            // Process quote asset
            this.processAsset({
                asset_id: pair.quote_asset_id,
                ticker: pair.quote_asset,
                name: pair.quote_asset,
                precision: pair.quote_precision,
                is_active: pair.is_active,
                min_order_size: pair.min_quote_order_size,
                max_order_size: pair.max_quote_order_size,
                trading_partner: pair.base_asset_id
            });
        });
    }
    /**
     * Process a single asset from pair data
     */
    processAsset(assetData) {
        const existing = this.assetMap.get(assetData.asset_id);
        if (existing) {
            // Add trading partner if not already present
            if (!existing.trading_pairs.includes(assetData.trading_partner)) {
                existing.trading_pairs.push(assetData.trading_partner);
            }
            // Update min/max order sizes (take most restrictive)
            existing.min_order_size = Math.max(existing.min_order_size, assetData.min_order_size);
            existing.max_order_size = Math.min(existing.max_order_size, assetData.max_order_size);
        }
        else {
            // Create new asset entry
            this.assetMap.set(assetData.asset_id, {
                asset_id: assetData.asset_id,
                ticker: assetData.ticker,
                name: assetData.name,
                precision: assetData.precision,
                is_active: assetData.is_active,
                min_order_size: assetData.min_order_size,
                max_order_size: assetData.max_order_size,
                trading_pairs: [assetData.trading_partner]
            });
        }
    }
    /**
     * Find asset by ticker
     */
    findByTicker(ticker) {
        for (const asset of this.assetMap.values()) {
            if (asset.ticker.toLowerCase() === ticker.toLowerCase()) {
                return asset;
            }
        }
        return undefined;
    }
    /**
     * Find asset by ID
     */
    findById(assetId) {
        return this.assetMap.get(assetId);
    }
    /**
     * Get all assets
     */
    getAllAssets() {
        return Array.from(this.assetMap.values());
    }
    /**
     * Check if two assets can be traded directly
     */
    canTrade(fromAssetId, toAssetId) {
        const fromAsset = this.assetMap.get(fromAssetId);
        return fromAsset ? fromAsset.trading_pairs.includes(toAssetId) : false;
    }
    /**
     * Get trading partners for an asset
     */
    getTradingPartners(assetId) {
        const asset = this.assetMap.get(assetId);
        if (!asset)
            return [];
        return asset.trading_pairs
            .map(partnerId => this.assetMap.get(partnerId))
            .filter((partner) => partner !== undefined);
    }
    /**
     * Get all active trading pairs
     */
    getActivePairs() {
        return this.pairs.filter(pair => pair.is_active);
    }
}
exports.AssetPairMapper = AssetPairMapper;
/**
 * Create an AssetPairMapper instance
 */
function createAssetPairMapper(pairsResponse) {
    return new AssetPairMapper(pairsResponse);
}
