// @kaleidoswap/kaleidoswap-sdk/dist/utils/retry.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createRetryWrapper = exports.withRetry = exports.defaultRetryConfig = void 0;
exports.retry = retry;
const exceptions_1 = require("../types/exceptions");
const errors_1 = require("../types/errors");
exports.defaultRetryConfig = {
    maxRetries: 3,
    initialDelay: 1000,
    maxDelay: 10000,
    exponentialBase: 2,
    jitter: true,
    retryOnExceptions: [exceptions_1.NetworkError, exceptions_1.RateLimitError],
    respectErrorRetryConfig: true
};
/**
 * Enhanced retry function that works with the new error system
 */
async function retry(fn, config = {}) {
    const finalConfig = { ...exports.defaultRetryConfig, ...config };
    let lastError;
    for (let attempt = 0; attempt <= finalConfig.maxRetries; attempt++) {
        try {
            return await fn();
        }
        catch (error) {
            lastError = error;
            // Enhanced retry logic for KaleidoSDKError
            if (error instanceof errors_1.KaleidoSDKError && finalConfig.respectErrorRetryConfig) {
                // Use error's own retry configuration if available
                if (!error.isRetryable()) {
                    throw error; // Don't retry non-retryable errors
                }
                // Use error's own retry limits if configured
                const errorMaxRetries = error.metadata.maxRetries;
                if (errorMaxRetries !== undefined && attempt >= errorMaxRetries) {
                    throw error;
                }
            }
            else {
                // Legacy retry logic for backward compatibility
                const shouldRetry = finalConfig.retryOnExceptions.some((ExceptionClass) => error instanceof ExceptionClass);
                if (!shouldRetry) {
                    throw error;
                }
            }
            // Don't retry if we've reached max attempts
            if (attempt === finalConfig.maxRetries) {
                throw error;
            }
            // Calculate delay based on error's retry strategy or config
            let delay = finalConfig.initialDelay;
            if (error instanceof errors_1.KaleidoSDKError && finalConfig.respectErrorRetryConfig) {
                const errorDelay = error.metadata.retryDelay;
                const retryStrategy = error.getRetryStrategy();
                if (errorDelay) {
                    delay = errorDelay;
                }
                else {
                    delay = calculateDelayFromStrategy(retryStrategy, attempt, finalConfig.initialDelay, finalConfig.exponentialBase, finalConfig.maxDelay);
                }
            }
            else {
                // Standard exponential backoff
                delay = Math.min(finalConfig.initialDelay * Math.pow(finalConfig.exponentialBase, attempt), finalConfig.maxDelay);
            }
            // Add jitter if enabled
            const finalDelay = finalConfig.jitter
                ? delay + Math.random() * delay * 0.1
                : delay;
            await new Promise(resolve => setTimeout(resolve, finalDelay));
        }
    }
    throw lastError;
}
/**
 * Calculate delay based on retry strategy
 */
function calculateDelayFromStrategy(strategy, attempt, initialDelay, exponentialBase, maxDelay) {
    switch (strategy) {
        case errors_1.RetryStrategy.IMMEDIATE_RETRY:
            return 0;
        case errors_1.RetryStrategy.LINEAR_BACKOFF:
            return Math.min(initialDelay * (attempt + 1), maxDelay);
        case errors_1.RetryStrategy.EXPONENTIAL_BACKOFF:
            return Math.min(initialDelay * Math.pow(exponentialBase, attempt), maxDelay);
        case errors_1.RetryStrategy.NO_RETRY:
        default:
            return 0; // This shouldn't be called for NO_RETRY, but just in case
    }
}
/**
 * Convenient wrapper for retry with predefined configuration
 */
const withRetry = (fn, config) => retry(fn, config);
exports.withRetry = withRetry;
/**
 * Creates a retry wrapper for a function
 */
const createRetryWrapper = (fn, config = {}) => {
    return async (...args) => {
        return retry(() => fn(...args), config);
    };
};
exports.createRetryWrapper = createRetryWrapper;
